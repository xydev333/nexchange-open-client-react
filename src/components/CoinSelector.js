import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { selectCoin, updateAmounts, fetchPrice } from '../actions/index.js';

class CoinSelector extends Component {

	constructor(props) {
		super(props);
		this.state = {
			selectedCoin: this.props.selectedCoin.present[this.props.type],
			isDropdownVisible: false
		}

    	this.selectCoin = this.selectCoin.bind(this);
	}

	selectCoin(coin) {
		this.props.selectedCoin.present[this.props.type] = coin;
		this.props.selectCoin(this.props.selectedCoin);

		let nextProps = Object.assign({}, this.props.amounts);
		nextProps['update'] = true;
		nextProps['lastEdited'] = 'deposit';
		this.props.updateAmounts(nextProps);

		this.props.fetchPrice(`${this.props.selectedCoin.present.deposit}${this.props.selectedCoin.present.receive}`);

		this.setState({
			isDropdownVisible: false,
			selectedCoin: coin
		});
	}

	componentWillReceiveProps(nextProps) {
		this.setState({
			selectedCoin: this.props.selectedCoin.present[this.props.type]
		});
	}

	render() {
		let coins = this.props.coinsInfo.map(coin => {
			return (
				<div className="row coin" key={coin.ticker} onClick={() => this.selectCoin(coin.ticker)}>
					<div className="col-xs-4">{coin.ticker}</div>
					<div className="col-xs-3 text-center">
						<i className={`coin-icon ${coin.ticker}`}></i>
					</div>
					<div className="col-xs-5">{coin.name}</div>
				</div>
			);
		});

		return (
			<div>
				<div className="selectedCoin" onClick={() => this.setState({isDropdownVisible: !this.state.isDropdownVisible})}>
					{this.state.selectedCoin} <i className={`coin-icon ${this.state.selectedCoin}`}></i>
				</div>

				{this.state.isDropdownVisible ? <div className="coin-currency-dropdown">{coins}</div> : null}
			</div>
		);
	}
}


function mapStateToProps(state) {
	return {
		selectedCoin: state.selectedCoin,
		coinsInfo: state.coinsInfo,
		amounts: state.amounts
	}
}

function mapDispatchToProps(dispatch) {
	return bindActionCreators({
		selectCoin: selectCoin,
		updateAmounts: updateAmounts,
		fetchPrice: fetchPrice
	}, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(CoinSelector);
