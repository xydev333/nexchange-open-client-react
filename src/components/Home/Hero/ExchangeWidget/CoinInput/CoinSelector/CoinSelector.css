.selected-coin {
  position: absolute;
  bottom: 21px;
  right: 25px;
  cursor: pointer;
  font-size: 18px;
  font-weight: 700;
  color: #fff;
}

.coin-icon.coin-icon {
  position: relative;
  top: 1px;
  color: #fff;
}

.arrow-down {
  position: relative;
  top: 2px;
  margin-left: 8px;
}

.span {
  position: relative;
  top: 2px;
  left: -10px;
}
