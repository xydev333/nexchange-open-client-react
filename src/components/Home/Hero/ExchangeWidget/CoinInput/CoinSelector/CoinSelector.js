import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import onClickOutside from 'react-onclickoutside';
import { selectCoin, fetchPrice, setWallet, errorAlert } from 'Actions/index.js';
import CoinsDropdown from './CoinsDropdown/CoinsDropdown';
import styles from './CoinSelector.css';

class CoinSelector extends Component {
  state = {
    isDropdownVisible: false,
    selectedCoin: null,
  };

  selectCoin = coin => {
    this.props.selectCoin({
      ...this.props.selectedCoin,
      [this.props.type]: coin,
      lastSelected: this.props.type,
    });

    this.setState({ isDropdownVisible: false });
    ga('send', 'event', 'Order', 'select coin');
  };

  handleClickOutside = event => {
    this.setState({ isDropdownVisible: false });
  };

  calculateDepositAmount = coin => {
    return ['EUR', 'GBP', 'USD'].indexOf(coin.name) > -1 ? 100 : parseFloat(coin.minimal_amount) * 100;
  };

  handleClick = code => {
    this.selectCoin(code);
    this.props.onSelect();
  };

  UNSAFE_componentWillReceiveProps = nextProps => {
    // This condition means that we have selected default currency pairs
    // and now need to fetch price.
    if (this.props.selectedCoin.deposit === null && nextProps.selectedCoin.deposit && nextProps.type === 'deposit') {
      this.props.fetchPrice({
        pair: `${nextProps.selectedCoin.receive}${nextProps.selectedCoin.deposit}`,
        lastEdited: 'deposit',
      });
    }

    // This condition means that selected coin has been changed and price
    // needs to be refetched.
    if (
      this.props.selectedCoin[this.props.type] !== nextProps.selectedCoin[this.props.type] &&
      ((this.props.type === 'deposit' && nextProps.selectedCoin.lastSelected === 'deposit') ||
        (this.props.type === 'receive' && nextProps.selectedCoin.lastSelected === 'receive'))
    ) {
      this.props.fetchPrice({
        pair: `${nextProps.selectedCoin.receive}${nextProps.selectedCoin.deposit}`,
        lastEdited: nextProps.selectedCoin.lastSelected,
      });
    }

    // Check if pair is valid. If not, show error.
    if (
      nextProps.selectedCoin.deposit &&
      nextProps.selectedCoin.receive &&
      !this.props.pairs[nextProps.selectedCoin.deposit][nextProps.selectedCoin.receive]
    ) {
      const validPairs = Object.keys(this.props.pairs[nextProps.selectedCoin.deposit])
        .map(coin => coin)
        .filter(coin => this.props.pairs[nextProps.selectedCoin.deposit][coin] === true)
        .join(', ');

      this.props.errorAlert({
        message: `You cannot buy ${nextProps.selectedCoin.receive} with ${nextProps.selectedCoin.deposit}. Try ${validPairs}.`,
        show: true,
        type: 'INVALID_PAIR',
      });
    }
  };

  render() {
    const type = this.props.type;
    const selectedCoin = this.props.selectedCoin[type];

    if (!selectedCoin) return null;

    return (
      <div>
        <div
          className={`selected-coin ${styles['selected-coin']}`}
          onClick={() => this.setState({ isDropdownVisible: !this.state.isDropdownVisible })}
        >
          <span className={styles.span}>{selectedCoin}</span>
          <i className={`${styles['coin-icon']} cc ${selectedCoin}`} />
          <i className={`fas fa-angle-down ${styles['arrow-down']}`} />
        </div>

        {this.state.isDropdownVisible && <CoinsDropdown type={type} onClick={this.handleClick} coinsInfo={this.props.coinsInfo} />}
      </div>
    );
  }
}

const mapStateToProps = ({ selectedCoin, coinsInfo, pairs, price }) => ({ selectedCoin, coinsInfo, pairs, price });
const mapDispatchToProps = dispatch => bindActionCreators({ selectCoin, fetchPrice, setWallet, errorAlert }, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(onClickOutside(CoinSelector));
