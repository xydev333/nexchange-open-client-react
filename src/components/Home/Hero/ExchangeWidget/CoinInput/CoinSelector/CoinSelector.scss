.selected-coin {
  position: absolute;
  bottom: 10px;
  left: 0;
  padding: 11px 25px 13px 15px;
  cursor: pointer;
  font-size: 18px;
  font-weight: 700;
  color: $black;
}

.carret {
  display: inline-block;
  position: relative;
  top: -4px;
  left: 8px;
  width: 8px;
  height: 5px;
  background-image: url('../../../../../../img/carret-down.svg');
}

.coin-icon.coin-icon {
  position: relative;
  top: 1px;
  margin-right: 12px;
  font-size: 24px;
}
