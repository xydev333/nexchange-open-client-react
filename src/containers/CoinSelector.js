import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import onClickOutside from 'react-onclickoutside';
import Helpers from '../helpers';

import { selectCoin, fetchPrice, setWallet } from '../actions/index.js';


class CoinSelector extends Component {
	constructor(props) {
		super(props);
		this.state = {
			isDropdownVisible: false,
			selectedCoin: null
		}

    this.selectCoin = this.selectCoin.bind(this);
	}

	selectCoin(coin) {
		this.props.selectCoin(Object.assign({}, this.props.selectedCoin, {[this.props.type]: coin, lastSelected: this.props.type}));

		this.setState({isDropdownVisible: false});

		ga('send', 'event', 'Order', 'select coin');
	}

	handleClickOutside(event) {
		this.setState({isDropdownVisible: false});
	}

	componentWillReceiveProps(nextProps) {
		if (this.props.selectedCoin[this.props.type] !== nextProps.selectedCoin[this.props.type]) {
			this.props.fetchPrice({pair: `${nextProps.selectedCoin.receive}${nextProps.selectedCoin.deposit}`, lastEdited: nextProps.amounts.lastEdited, amount: nextProps.amounts[nextProps.amounts.lastEdited]});
		}
	}

	render() {
		let selectedCoin = this.props.selectedCoin[this.props.type],
			type = this.props.type,
			lastSelectedType = this.props.selectedCoin.lastSelected,
			lastSelectedCoin = this.props.selectedCoin[lastSelectedType],
			filteredCoins = this.props.coinsInfo.filter(coin => {
	   		let params = Helpers.urlParams();

				if (this.props.pairs && type !== lastSelectedType && lastSelectedCoin !== coin.code) {
					if (lastSelectedType === 'deposit') {
						return this.props.pairs[lastSelectedCoin][coin.code];
					} else if (lastSelectedType === 'receive') {
						return this.props.pairs[coin.code][lastSelectedCoin];
					}
				}

	      if (params && params.hasOwnProperty('test')) {
					return (type.toUpperCase() === 'DEPOSIT') ? coin.is_quote_of_enabled_pair_for_test : coin.is_base_of_enabled_pair_for_test;
				}

				return (type.toUpperCase() === 'DEPOSIT') ? coin.is_quote_of_enabled_pair : coin.is_base_of_enabled_pair;
			}),
			coins = filteredCoins.map(coin => {
				return (
					<div className="row coin" key={coin.code} onClick={() => this.selectCoin(coin.code)}>
						<div className="col-xs-4">{coin.code}</div>
						<div className="col-xs-3 text-center">
							<i className={`cc-${coin.code} ${coin.code}`}></i>
						</div>
						<div className="col-xs-5 text-capitalize">{coin.name}</div>
					</div>
				);
			});

		return (
			<div>
				<div className="selectedCoin" onClick={() => this.setState({isDropdownVisible: !this.state.isDropdownVisible})}>
					<span>{selectedCoin}</span>
					<i className={`coin-icon cc-${selectedCoin}`}></i>
					<i className="fa fa-angle-down"></i>
				</div>

				{this.state.isDropdownVisible ? <div className="coin-currency-dropdown">{coins}</div> : null}
			</div>
		);
	}
}


function mapStateToProps(state) {
	return {
		selectedCoin: state.selectedCoin,
		coinsInfo: state.coinsInfo,
		amounts: state.amounts,
		pairs: state.pairs
	}
}

function mapDispatchToProps(dispatch) {
	return bindActionCreators({
		selectCoin: selectCoin,
		fetchPrice: fetchPrice,
		setWallet: setWallet,
	}, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(onClickOutside(CoinSelector));
