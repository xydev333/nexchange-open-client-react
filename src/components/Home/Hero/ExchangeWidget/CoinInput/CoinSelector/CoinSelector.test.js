import React from 'react';
import { shallow, mount } from 'enzyme';
import CoinSelector from './CoinSelector.js';
import coinsInfo from 'Mocks/currency.js';

describe('CoinSelector', () => {
  let wrapShallowDeposit, wrapShallowReceive, wrapMountDeposit;

  beforeEach(() => {
    wrapShallowDeposit = shallow(<CoinSelector type="deposit" onClick={jest.fn()} coinsInfo={coinsInfo} />);
    wrapShallowReceive = shallow(<CoinSelector type="receive" onClick={jest.fn()} coinsInfo={coinsInfo} />);
    wrapMountDeposit = mount(<CoinSelector type="deposit" onClick={jest.fn()} coinsInfo={coinsInfo} />);
  });

  it('renders correctly', () => {
    expect(wrapShallowDeposit).toMatchSnapshot();
    expect(wrapShallowReceive).toMatchSnapshot();
  });

  it('deposit dropdown contains correct coins', () => {
    for (const coin of coinsInfo) {
      if (coin.is_quote_of_enabled_pair) {
        expect(wrapShallowDeposit.find(`[data-test="${coin.code}"]`).length).toEqual(1);
      } else {
        expect(wrapShallowDeposit.find(`[data-test="${coin.code}"]`).length).toEqual(0);
      }
    }
  });

  it('receive dropdown contains correct coins', () => {
    for (const coin of coinsInfo) {
      if (coin.is_base_of_enabled_pair) {
        expect(wrapShallowReceive.find(`[data-test="${coin.code}"]`).length).toEqual(1);
      } else {
        expect(wrapShallowReceive.find(`[data-test="${coin.code}"]`).length).toEqual(0);
      }
    }
  });

  it('search input gets correct value after state change and filters coins', () => {
    wrapMountDeposit.setState({
      value: 'bit',
    });

    const input = wrapMountDeposit.find(`[data-test="search"]`);
    expect(input.props().value).toEqual('bit');

    expect(wrapMountDeposit.find(`[data-test="BTC"]`).length).toEqual(1);
    expect(wrapMountDeposit.find(`[data-test="BCH"]`).length).toEqual(1);
    expect(wrapMountDeposit.find(`.coin`).length).toEqual(2);
  });
});
