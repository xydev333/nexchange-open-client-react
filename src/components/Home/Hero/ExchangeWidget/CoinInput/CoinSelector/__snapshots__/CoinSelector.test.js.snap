// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinSelector renders correctly 1`] = `undefined`;

exports[`CoinSelector renders correctly 2`] = `undefined`;
